File: android/src/main/java/expo/modules/mediapipepose/PoseMaskStore.kt
package expo.modules.mediapipepose

import android.content.Context
import android.graphics.Bitmap
import android.util.Log
import com.google.mediapipe.framework.image.ByteBufferExtractor
import com.google.mediapipe.framework.image.MPImage
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarkerResult
import java.io.File
import java.nio.ByteOrder
import java.util.UUID
import java.util.concurrent.CompletableFuture
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.math.roundToInt

internal class PoseMaskStore {
    companion object {
        private val startupLock = Any()
        private var startupCleaned = false
        private const val directoryPrefix = "expo-mediapipe-pose-masks-"

        private fun cleanPreviousProcessFiles(cache: File) =
            synchronized(startupLock) {
                if (startupCleaned) return@synchronized
                val entries = requireNotNull(cache.listFiles()) { "Cannot inspect mask cache" }
                for (entry in entries) {
                    if (!entry.name.startsWith(directoryPrefix)) continue
                    check(entry.deleteRecursively()) { "Cannot clear previous mask files" }
                }
                startupCleaned = true
            }
    }

    private val directoryName = "$directoryPrefix${UUID.randomUUID()}"
    private var directory: File? = null
    private val leases = mutableMapOf<String, File>()
    private val closed = AtomicBoolean(false)

    @Synchronized
    fun save(
        context: Context,
        result: PoseLandmarkerResult,
        width: Int,
        height: Int,
        maximumDimension: Int,
    ): Map<String, Any> {
        check(!closed.get()) { "Mask owner is closed" }
        if (result.landmarks().isEmpty()) return mapOf("status" to "empty")
        if (leases.size >= 2) return mapOf("status" to "backpressure")
        val masks =
            result.segmentationMasks().orElseThrow {
                IllegalStateException("Segmentation masks are missing")
            }
        check(masks.size == result.landmarks().size)
        val id = UUID.randomUUID().toString()
        cleanPreviousProcessFiles(context.cacheDir)
        val root = File(context.cacheDir, directoryName)
        directory = root
        val folder = File(root, id)
        check(folder.mkdirs()) { "Cannot create mask directory" }
        leases[id] = folder
        try {
            val payload = masks.mapIndexed { index, mask ->
                val file = File(folder, "$index.png")
                val size = encode(mask, file, maximumDimension)
                mapOf(
                    "poseIndex" to index,
                    "uri" to file.toURI().toString(),
                    "width" to size.first,
                    "height" to size.second,
                )
            }
            return mapOf(
                "status" to "available",
                "leaseId" to id,
                "masks" to payload,
                "imageSize" to mapOf("width" to width, "height" to height),
            )
        } catch (error: Exception) {
            release(id)
            throw error
        }
    }

    @Synchronized
    fun release(id: String) {
        val folder = leases[id] ?: return
        check(folder.deleteRecursively()) { "Cannot remove mask output" }
        leases.remove(id)
        folder.parentFile?.delete()
    }

    fun discard(segmentation: Map<String, Any>?) {
        val id = segmentation?.get("leaseId") as? String ?: return
        CompletableFuture.runAsync {
            try {
                release(id)
            } catch (error: Exception) {
                Log.e("PoseMasks", "Mask cleanup failed", error)
            }
        }
    }

    fun destroy() {
        closed.set(true)
        CompletableFuture.runAsync {
            synchronized(this) {
                val root = directory
                if (root != null && !root.deleteRecursively())
                    Log.e("PoseMasks", "Mask cleanup failed")
                leases.clear()
            }
        }
    }

    private fun encode(mask: MPImage, file: File, maximumDimension: Int): Pair<Int, Int> {
        require(mask.width > 0 && mask.height > 0)
        val scale = minOf(1.0, maximumDimension.toDouble() / maxOf(mask.width, mask.height))
        val width = maxOf(1, (mask.width * scale).toInt())
        val height = maxOf(1, (mask.height * scale).toInt())
        val probabilities =
            ByteBufferExtractor.extract(mask).order(ByteOrder.nativeOrder()).asFloatBuffer()
        require(probabilities.remaining() == mask.width * mask.height)
        val pixels = IntArray(width * height)
        for (row in 0 until height) {
            val sourceRow = minOf(mask.height - 1, ((row + 0.5) * mask.height / height).toInt())
            for (column in 0 until width) {
                val sourceColumn =
                    minOf(mask.width - 1, ((column + 0.5) * mask.width / width).toInt())
                val probability = probabilities.get(sourceRow * mask.width + sourceColumn)
                require(probability.isFinite())
                val alpha = (probability.coerceIn(0f, 1f) * 255).roundToInt()
                pixels[row * width + column] = (alpha shl 24) or 0x00ffffff
            }
        }
        val bitmap = Bitmap.createBitmap(pixels, width, height, Bitmap.Config.ARGB_8888)
        try {
            file.outputStream().use { output ->
                check(bitmap.compress(Bitmap.CompressFormat.PNG, 100, output))
            }
        } finally {
            bitmap.recycle()
        }
        return width to height
    }
}

File: ios/PoseMaskStore.swift
import Foundation
import ImageIO
import MediaPipeTasksVision
import UniformTypeIdentifiers

internal final class PoseMaskStore {
  private static let startupLock = NSLock()
  private static var startupCleaned = false
  private static let directoryPrefix = "expo-mediapipe-pose-masks-"
  private let worker = DispatchQueue(label: "expo.pose.masks")
  private let directoryName = "\(PoseMaskStore.directoryPrefix)\(UUID().uuidString)"
  private var directory: URL?
  private var leases: [String: URL] = [:]
  private var closed = false

  func save(_ result: PoseLandmarkerResult, width: Int, height: Int, maximumDimension: Int) throws
    -> [String: Any]
  {
    try worker.sync {
      guard !closed else { throw PoseMediaError.maskUnavailable }
      if result.landmarks.isEmpty { return ["status": "empty"] }
      guard leases.count < 2 else { return ["status": "backpressure"] }
      let masks = result.segmentationMasks
      guard masks.count == result.landmarks.count else { throw PoseMediaError.maskUnavailable }
      let cache = try FileManager.default.url(
        for: .cachesDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
      try Self.cleanPreviousProcessFiles(in: cache)
      let root = cache.appendingPathComponent(directoryName)
      directory = root
      let id = UUID().uuidString
      let folder = root.appendingPathComponent(id)
      try FileManager.default.createDirectory(at: folder, withIntermediateDirectories: true)
      leases[id] = folder
      do {
        let payload = try masks.enumerated().map { index, mask -> [String: Any] in
          let url = folder.appendingPathComponent("\(index).png")
          let size = try encode(mask, to: url, maximumDimension: maximumDimension)
          return [
            "poseIndex": index, "uri": url.absoluteString, "width": size.width,
            "height": size.height,
          ]
        }
        return [
          "status": "available", "leaseId": id, "masks": payload,
          "imageSize": ["width": width, "height": height],
        ]
      } catch {
        try remove(id)
        throw error
      }
    }
  }
  func release(_ id: String) throws { try worker.sync { try remove(id) } }
  func discard(_ segmentation: [String: Any]?) {
    guard let id = segmentation?["leaseId"] as? String else { return }
    worker.async {
      do { try self.remove(id) } catch {
        NSLog("Pose mask cleanup failed: %@", String(describing: error))
      }
    }
  }
  func destroy() {
    worker.async {
      self.closed = true
      guard let directory = self.directory else { return }
      guard FileManager.default.fileExists(atPath: directory.path) else {
        self.leases.removeAll()
        return
      }
      do {
        try FileManager.default.removeItem(at: directory)
        self.leases.removeAll()
      } catch { NSLog("Pose mask cleanup failed: %@", String(describing: error)) }
    }
  }
  private static func cleanPreviousProcessFiles(in cache: URL) throws {
    startupLock.lock()
    defer { startupLock.unlock() }
    guard !startupCleaned else { return }
    for directory in try FileManager.default.contentsOfDirectory(
      at: cache, includingPropertiesForKeys: nil)
    {
      if directory.lastPathComponent.hasPrefix(directoryPrefix) {
        try FileManager.default.removeItem(at: directory)
      }
    }
    startupCleaned = true
  }
  private func remove(_ id: String) throws {
    guard let folder = leases[id] else { return }
    if FileManager.default.fileExists(atPath: folder.path) {
      try FileManager.default.removeItem(at: folder)
    }
    leases.removeValue(forKey: id)
  }
  private func encode(_ mask: Mask, to url: URL, maximumDimension: Int) throws -> (
    width: Int, height: Int
  ) {
    guard mask.width > 0, mask.height > 0 else { throw PoseMediaError.maskUnavailable }
    let scale = min(1, Double(maximumDimension) / Double(max(mask.width, mask.height)))
    let width = max(1, Int(Double(mask.width) * scale))
    let height = max(1, Int(Double(mask.height) * scale))
    let probabilities = mask.float32Data
    var pixels = [UInt8](repeating: 255, count: width * height * 4)
    for row in 0..<height {
      let sourceRow = min(
        mask.height - 1, Int((Double(row) + 0.5) * Double(mask.height) / Double(height)))
      for column in 0..<width {
        let sourceColumn = min(
          mask.width - 1, Int((Double(column) + 0.5) * Double(mask.width) / Double(width)))
        let probability = probabilities[sourceRow * mask.width + sourceColumn]
        guard probability.isFinite else { throw PoseMediaError.maskUnavailable }
        pixels[(row * width + column) * 4 + 3] = UInt8(
          (min(1, max(0, probability)) * 255).rounded())
      }
    }
    let data = Data(pixels)
    guard let provider = CGDataProvider(data: data as CFData),
      let image = CGImage(
        width: width, height: height, bitsPerComponent: 8, bitsPerPixel: 32, bytesPerRow: width * 4,
        space: CGColorSpaceCreateDeviceRGB(),
        bitmapInfo: CGBitmapInfo(rawValue: CGImageAlphaInfo.last.rawValue), provider: provider,
        decode: nil, shouldInterpolate: false, intent: .defaultIntent),
      let destination = CGImageDestinationCreateWithURL(
        url as CFURL, UTType.png.identifier as CFString, 1, nil)
    else { throw PoseMediaError.maskUnavailable }
    CGImageDestinationAddImage(destination, image, nil)
    guard CGImageDestinationFinalize(destination) else { throw PoseMediaError.maskUnavailable }
    return (width, height)
  }
}
